File: datastructures/linkedlist/doublyLinkedList.js
class Node {
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class DoublyLinkedList {
    constructor() {
        this.head = null;
        this.printString = "";
        this.reversePrintString = "";

        this.insertAtHead = this.insertAtHead.bind(this);
        this.printList = this.printList.bind(this);
        this.reversePrint = this.reversePrint.bind(this);
    }

    insertAtHead(d) {
        var newNode = new Node(d);
        if (!this.head) {
            this.head = newNode;
            return;
        }

        this.head.prev = newNode;
        newNode.next = this.head;
        this.head = newNode;
    }

    printList() {
        var temp = this.head;
        this.printString = "Printing list: ";
        while (temp) {
            this.printString += ("[" + temp.data + "] ");
            temp = temp.next;
        }
        this.printString += ".";
        console.log(this.printString);
    }

    reversePrint() {
        var temp = this.head;
        if (!temp) return;

        while (temp.next) {
            temp = temp.next;
        }

        this.reversePrintString = "Reverse printing: ";
        while (temp) {
            this.reversePrintString += ("[" + temp.data + "] ");
            temp = temp.prev;
        }
        this.reversePrintString += ".";
        console.log(this.reversePrintString);
    }
}

console.log('Starting job...');

var myList = new DoublyLinkedList();
myList.insertAtHead(6);
myList.insertAtHead(4);
myList.insertAtHead(2);

myList.printList();
myList.reversePrint();

console.log('Job finished.');

File: datastructures/linkedlist/doublyLinkedList.java
class DoublyLinkedList {
    Node head;

    class Node {
        int data;
        Node next;
        Node prev;

        Node(int d) { data = d; next = null; prev = null; };
    }

    public void insertAtHead(int d) {
        Node newNode = new Node(d);
        if (head == null) {
            head = newNode;
            return;
        }
        head.prev = newNode;
        newNode.next = head;
        head = newNode;
    }

    public void print() {
        Node temp = head;
        System.out.print("The list is: ");
        while (temp != null) {
            System.out.print("[" + temp.data + "] ");
            temp = temp.next;
        }
        System.out.println();
    }

    public void reversePrint() {
        Node temp = head;
        if (temp == null) {
            return;
        }

        while (temp.next != null) {
            temp = temp.next;
        }

        System.out.print("Reverse printing: ");
        while (temp != null) {
            System.out.print("[" + temp.data + "] ");
            temp = temp.prev;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        DoublyLinkedList myList = new DoublyLinkedList();
        
        myList.insertAtHead(6);
        myList.insertAtHead(4);
        myList.insertAtHead(2);

        myList.print();
        myList.reversePrint();
    }
}